2/answers.txt

```diff
@@ -0,0 +1,231 @@
+Multiple processes cannot share main memory space however, and the creation of share memory segments is a relatively resource expensive task, and should only be used if it would create a larger benefit.
+
+Q21: What are threads? Can they share memory?
+
+Threads are the lighter brother of a process, they are single streams of code execution occuring within the heap, a single thread process will have just one, a multi thread process will obviously have many.
+Each thread has its own stack, but share the heap for the process entire, meaning there is inherently some shared nature of memory within the heap, but the stack itself is isolated.
+
+Q22: What is CPU IPC and frequency? Why is CPU speed more important for single threaded applications than multi-threaded applications?
+
+Instructions per clock or cycle. CPU speed is more important for a single threaded application because that application will only be functioning on a single core, as a result the speed of that core
+will dictate the speed of that program. With multithreaded applications, each thread can be split between cores in a type of load balancing, making the overall impact of core speed less for that application.
+
+Q23: What is the difference between concurrent and parallel execution? What does 'thread safety' mean?
+
+Concurrent execution is when multiple processes are being worked upon at the same time, but not executed at the same time, whereas parallel execution is where 2 or more processes are being executed simultaneously.
+An application can be any number of the above, i.e. concurrent but not parallel, parallel but not concurrent, neither and both. Thread safety applies specifically to multi threaded code. 
+
+Thread safety is about ensuring that data structure manipulated by multiple threads affect the data in the intended way, without causing unexpected changes or damage as a result of multiple entry points to the data structure.
+Typically these are done by implementing race conditions, however you can achieve thread safety without race conditions by implementing, Re-Entrancy, thread-local storage, immutable objects, mutual exclusion and atomic operations.
+
+Q24: If a CPU has 4 cores, how many processes can it execute concurrently?
+
+A CPU with 4 cores can run as many processes "Concurrently" as it has memory to store the not currently executing processes.
+
+Q25: Explain the difference between synchronous and asynchronous code. When would you prefer one technique over the other?
+
+Synchronous code has to be sequentially completed before the next set of instructions can be processed, asynchronous code allow an instruction set to be hung and another started and then returned back to the previous and so on.
+
+You would use asynchronous techniques where you can run two tasks simultaneously, usually when they do not share dependencies and are independant, and you would use synchronous techniques, when you need a task to complete before proceeding.
```

Comment:
Typically, if it can be executed synchronously, without downsides, it should be executed synchronously.
Synchronous code is easier to write and to reason about than async code.
Synchronous code is also (marginally) more performant than async code.

Use async when you encounter something that blocks your code for a non-trivial period of time (eg IO) and you wish to be able to execute other tasks during the time that you would otherwise be waiting.